### ISSUE TITLE
Missing rules for cards

### ISSUE THREAD

**Issue by jsterner73** [KFNEXUS/KFNEXUS.github.io#4]
I recently discovered that there's more to the Fleet Builder than the card data,

Actually there is basically logic behind each and every card that changes the cost of cards or adds slots to the ship.

So, I'm starting this issue now, so that we can collect the cards that need (more) logic implemented.
- [ ] Khan Singh (The Classic Movies 1)
- [x] Standby Battle Stations (The Doomsday Machine)
- [x] Vrax (Wave 23: I.R.W. Jazkal)
- [x] Reman Bodyguards (Wave 23: I.R.W. Jazkal)
- [x] Nijil (Wave 23: I.R.W. Jazkal)
- [x] Prototype Cloaking Device (Wave 23: I.R.W. Jazkal)
- [x] Klingon Helmsman (Wave 23: I.K.S. Amar)
- [x] Klingon Navigator (Wave 23: I.K.S. Amar)
- [x] Stand By Torpedoes (Wave 23: I.K.S. Amar)
- [ ] Tebok (The Void OP: I.R.W. T'Met)
- [x] Charging Weapons (The Void OP: I.R.W. T'Met)
- [x] Self Repair Technology (The Void OP: I.R.W. T'Met)
- [x] Arming Sequence (Wave 22.5: Xindi Weapon Zero)
- [x] Degra (Wave 22.5: Xindi Weapon Zero)
- [x] Destructive Blast (Wave 22.5: Xindi Weapon Zero)
- [x] Rotating Emitters (Wave 22.5: Xindi Weapon Zero)
- [x] Subspace Vortex (Wave 22.5: Xindi Weapon Zero)
- [x] Self-Destruct (Wave 22.5: Xindi Weapon Zero)
- [ ] Covert Mission (Wave 22: R.I.S. Pi)
- [x] Distress Signal (Wave 22: R.I.S. Pi)
- [x] Vosk (Temporal Cold War: Temporal Cold War Cards)
- [x] Temporal Conduit (Temporal Cold War: Temporal Cold War Cards)
- [x] Photon Torpedoes (Temporal Cold War: Xindus)
- [x] Warp Drive Refit (Wave 5: Interceptor Five)

**Comment by KFNEXUS** [comment]
Collect all cards that needs logic is a good idea!
Khan Singh has a very complex logic(((

**Comment by jsterner73** [comment]
I think I got everything now - Azati Prime was the last ship to have special rules, After that I tried to find every potential card with rules to be implemented.

I will make this my priority for now and will return to spellchecking the cards afterwards.

**Comment by jsterner73** [comment]
I got most of the rules now - even Khan. The ones still missing take more to process, possibly I'll look into that later this week.

**Comment by CrazyVulcan** [comment]
Great job, its looking really good. Sorry you had to do the Reliant Prize cards
as I was watching startrek.com for their announcement and completely missed
that they were posted on the wizkids site.

**Comment by jsterner73** [comment]
No worries. :)

**Comment by KFNEXUS** [comment]
You rock! But i think khan slots this is not 3 new slots for ship, but configuration of slots that ship already has.

**Comment by jsterner73** [comment]
You are totally right, I misinterpretated the card effect. I'll see how to change that.

**Comment by CrazyVulcan** [comment]
Quick question on Wave 24, since it is just a re release of the original
Enterprise with as far as I can tell no changes to the cards do we feel
that we need to include a update for it? Or I suggest that we simply change
the original slot to "Wave 0 & 24: USS Enterprise" and make it easy on
ourselves?

**Comment by KFNEXUS** [comment]
> Or I suggest that we simply change
the original slot to "Wave 0 & 24: USS Enterprise" and make it easy on
ourselves?

I think there is a good solution

**Comment by jsterner73** [comment]
Although I like the idea of it being easy for us, I also like to see each item (in the set viewer) in chronological order. I'm not sure what that means for the fleet builder though - I guess a duplicate of everything with no reference to the set it came from.

I'll give it a try and see how it looks locally on my machine and then we can decide. Should be easy enough as we just need to copy every single card from Wave 0.

I'll let you know how it turned out, before we can make the decision.

**Comment by jsterner73** [comment]
Ok, there are no problems with duplicates in the fleet builder. I made a version with the name "U.S.S. Enterprise (repaint)" and the export function lists the cards like this, while only one "Christopher Pike" is available for select in the builder.

> U.S.S. Enterprise [22]
> - Christopher Pike 6 (Captain) (U.S.S. Enterprise, U.S.S. Enterprise (repaint)) [4]
> Ship Total: 26 SP
> 
> Fleet Total: 26 SP
> 
> Generated by Utopia
> http://kfnexus.github.io/staw-utopia/

I think we can do it this way, as it was easy enough to do and still has the chronological order in the set viewer.

**Comment by KFNEXUS** [comment]
> Ok, there are no problems with duplicates in the fleet builder. I made a version with the name "U.S.S. Enterprise (repaint)" and the export function lists the cards like this, while only one "Christopher Pike" is available for select in the builder.

Great news!

**Comment by CrazyVulcan** [comment]
It is a excellent solution.

**Comment by CrazyVulcan** [comment]
James T Kirk 8 from HMS Bounty needs scripting to reduce all Federation crew under him to cost -1SP

**Comment by jsterner73** [comment]
James T. Kirk (H.M.S. Bounty) scripting added.

**Comment by CrazyVulcan** [comment]
IRW Rateg, Main Batteries needs logic to add 1 Weapon Upgrade slot

**Comment by jsterner73** [comment]
Scripting added for Control Central (one per ship) and Main Batteries (one per ship, add weapon slot).

**Comment by CrazyVulcan** [comment]
Fleet Commander still needs JS work. I have exhausted my knowledge and ability trying to get it working. I can get it to add an extra slot, or add an addition captain slot next to the resource. But I have yet managed to get an extra captain above.

**Comment by jsterner73** [comment]
Where/how did you make your changes? I don't find anything in the rules file. On another note I think it's very difficult to make it work, if it's even possible at all given that we all don't have deeper knowledge about all the scrips and the structures of the program itself.

**Comment by CrazyVulcan** [comment]
I was playing around on my github.io since I did not wish to accidently screw up this build. I was trying to duplicate the data for the "Officer" resource. In an attempt to reproduce it. I know resources like the "Flagship", "Officer Exchange" and "Fleet Captain" add a new class that you then select from. It works kind of like Admirals. I was trying to change it so instead of the Class "Fleet Captain" it was using Captains as its class for the slot. 

https://crazyvulcan.github.io/staw-utopia/index.html#eyJzaGlwcyI6W3siaWQiOiJzaGlwOmZlZGVyYXRpb25fc3RhcnNoaXBfNzE4MDIiLCJjYXB0YWluIjp7ImlkIjoiY2FwdGFpbjpiZW5qYW1pbl9tYXh3ZWxsXzcyMDExIiwidXBncmFkZVNsb3RzIjpbe31dfSwidXBncmFkZXMiOlt7fSx7fSx7fSx7fSx7fV19LHsiaWQiOiJzaGlwOmZlZGVyYXRpb25fc3RhcnNoaXBfNzIwMDgiLCJjYXB0YWluIjp7ImlkIjoiY2FwdGFpbjp3aWxsaWFtX3Jvc3NfY2FwXzcyMDAxcCIsInVwZ3JhZGVTbG90cyI6W3t9XX0sInVwZ3JhZGVzIjpbe30se30se31dfV0sInJlc291cmNlIjp7ImlkIjoicmVzb3VyY2U6ZmxlZXRfY29tbWFuZGVyIiwidXBncmFkZVNsb3RzIjpbe31dfX0=

As you see here I have the data.json set up to add a new slot for all the ships, and it also added a class to the side of the resource. Ignore the fact that it says crewman, as I was trying to follow the bread crumbs as it were. You can easily change it to Captain. But the new spaces on the ships is where I would like this upgrade to be for a second Captain. 

I know I am on the right track I just need to get that last piece to click, I hope

**Comment by CrazyVulcan** [comment]
```
  "resources"` : [
		{
			"type" : "resource",
			"id" : "fleet_commander",
			"set" : ["72280r"],
			"name" : "Fleet Commander",
			"text" : "One ship in your fleet may have a 2nd Captain assigned to it. That ship may use both of its Captain's Abilitys as normal throughout the game.\n\nThe overall Captain Skill Number of the ship is equal to the highest Captain Skill of its Two Captains plus 1. \n\nIn addition, the ship to which the 2nd Captain is assigned gains +1 to either its Shield or Hull value (max 6 Hull). This decision must be made at the start of the game during the Gather Forces step and cannot be changed for the remainder of the game.",
			"cost" : 5,
                        "skill" : 1,
			"upgradeSlots" : [
				{
					"type" : ["captain"],
					"source" : "Add 2nd Captain"
				}
			],
			"showType" : true,
			"canEquip" : true,
			"intercept" :
			{
				"ship" : {},
				"fleet" : {}
			},
			"factionPenalty" : 0,
			"unique" : true,
			"showShipResourceSlot" : true
			},
```

**Comment by jsterner73** [comment]
I don't think you can do it in the data.json file. The rules logic is implemented in "\src\js\common\utopia-card-rules.js" and you need the tool _grunt_ to compile the files together Take a look at this file, there are some resources handled there already, perhaps you'll find some similarities there.

**Comment by CrazyVulcan** [comment]
Just downloaded the src folder to my git.io I will take a look over tomorrow as it is 3AM in the states and I need to sleep but I am optimistic.

**Comment by CrazyVulcan** [comment]
So far the best work around I have found to create a 'Captain' called Fleet Commander and have it so when you drop it into a captain slot in then generates 2 more captains and add's +1 to each of their Captain Skills. This does completely side skirts the need to have the Resource equipped but i want to work to fix that later. 

Right now the file comes in 2 parts first is a 'Captain' called 'Fleet Commander' who right now is a lv1 0SP captain. shown here.

```
{
			"type" : "captain",
			"id" : "fleet_commander_upgrade",
			"set" :["72280r"],
			"name": "Fleet Commander",
			"text" : "Add Two Captain Slots, Add +1 to highest Captain Skill",
			"cost" : 0,
			"skill" : 1,
			"factions" : ["independent"],
			"intercept" : {
				"ship" : {},
				"fleet" : {}
			}
		}
```
Then in the src/js/common file I have this.
```
		"fleet-commander:fleet_commander_upgrade":{
			//Add 2 new captain slots
			upgradeSlots: [
				{
					type: ["captain"];
					type: ["captain"]
				}
			];
			//No faction penalties
			intercept: {
				cost: function() { return 3; }
				ship: {
					skill: function(card,ship,fleet,skill) {
						if( card == ship.captain )
							return resolve(card,ship,fleet,skill) ? 1 );
						return skill;
					}
				}
		}
		},
```
Still trying to get the Javascript to work as I wish but as I have said this type of codeing is rather new to me. I am learning, I got this far with it after all. 

ps actually now that I am thinking of it I do not believe I have compiled the js file. hmm I will look into it but this is so you can see where I am with it.

**Comment by CrazyVulcan** [comment]
The New Enterprise-D has two cards that need scripting. Picard needs to reduce the cost of his ship by 2SP. He also can subtract 1SP for each upgrade assigned to him but he can not save more than 5SP total. Natasha Yarr needs to add 2 weapon slots. And Aft Phaser emitters needs to be scripted to change the attack value to be equal to the ship it is on's weapon value -1. It also can not be on a ship with a hull smaller than 4 and its cost will also be equal the ships weapon value. 

The Resource on the other hand..... I don't even have a guess as to how it can be implemented. its text is as such

Senior Staff -?SP
Each ship in your fleet may field 1 additional talent Upgrade for every crew Upgrade assigned to it. Each talent Upgrade is placed next to the crew Upgrade that fielded it. If the crew Upgrade is ever discarded, you must also discard the talent Upgrade that was fielded by that crew Upgrade. The cost of this resource is 1 sp for each additional talent Upgrade you assign to your fleet.

**Comment by jsterner73** [comment]
I added the scripting for Natasha Yar, Aft Phaser Emitters and Transporter. I also started on Jean-Luc Picard, but the upgrades are a bit tricky.

I got a Jekyll Page build error from GitHub, but I have no idea what that means.

**Comment by jsterner73** [comment]
I got the page to rebuild correctly with an empty commit, so the changes are online.
